tinyiiod.h

```diff
@@ -22,44 +22,55 @@
 #include <stddef.h>
 #include <stdint.h>
 #include <stdlib.h>
+#include <sys/types.h>
 
 #if !defined(__ssize_t_defined) && !defined(_SSIZE_T_DEFINED)
-typedef long int ssize_t;
 #define __ssize_t_defined
 #define _SSIZE_T_DEFINED
 #endif
 
 struct tinyiiod;
 
 struct tinyiiod_ops {
+	int32_t (*read)(int32_t *instance_id, char *buf, size_t len);
 	/* Read from the input stream */
-	void (*read)(char *buf, size_t len);
+	int32_t (*read_line)(int32_t *instance_id, char *buf);
+	int32_t (*read_nonbloking)(int32_t *instance_id, char *buf, size_t len);
+	int32_t (*read_wait)(int32_t *instance_id, size_t len);
 
 	/* Write to the output stream */
-	void (*write)(const char *buf, size_t len);
+	void (*write)(int32_t instance_id, const char *buf, size_t len);
+	int32_t (*exit)(int32_t instance_id);
 
 	ssize_t (*read_attr)(const char *device, const char *attr,
-			char *buf, size_t len, bool debug);
+			     char *buf, size_t len, bool debug);
 	ssize_t (*write_attr)(const char *device, const char *attr,
-			const char *buf, size_t len, bool debug);
+			      const char *buf, size_t len, bool debug);
 
 	ssize_t (*ch_read_attr)(const char *device, const char *channel,
-			bool ch_out, const char *attr, char *buf, size_t len);
+				bool ch_out, const char *attr, char *buf, size_t len);
 	ssize_t (*ch_write_attr)(const char *device, const char *channel,
-			bool ch_out, const char *attr,
-			const char *buf, size_t len);
+				 bool ch_out, const char *attr,
+				 const char *buf, size_t len);
 
-	int (*open)(const char *device, size_t sample_size, uint32_t mask);
-	int (*close)(const char *device);
+	int32_t (*open)(const char *device, size_t sample_size, uint32_t mask);
+	int32_t (*close)(const char *device);
 
-	ssize_t (*read_data)(const char *device, char *buf, size_t bytes_count);
+	ssize_t (*read_device)(const char *device, char **pbuf, size_t bytes_count);
+	ssize_t (*write_device)(const char *device, const char *buf,
+				size_t bytes_count);
+	int32_t (*read_reg)(uint32_t addr, int32_t *value);
+	int32_t (*write_reg)(uint32_t addr, uint32_t value);
 
-	int (*get_mask)(const char *device, uint32_t *mask);
+	int32_t (*get_mask)(const char *device, uint32_t *mask);
 };
 
 struct tinyiiod * tinyiiod_create(const char *xml,
-		const struct tinyiiod_ops *ops);
+				  const struct tinyiiod_ops *ops);
 void tinyiiod_destroy(struct tinyiiod *iiod);
-int tinyiiod_read_command(struct tinyiiod *iiod);
+int32_t tinyiiod_read_command(struct tinyiiod *iiod);
+int32_t tinyiiod_do_writebuf(struct tinyiiod *iiod, const char *device,
```

Comment:
the way `tinyiiod_read_command()` and `tinyiiod_do_writebuf()`, are structured, they don't look like API-style functions ; 
maybe they belong in ` tinyiiod-private.h` ?